equinox/jit.py

```diff
@@ -94,7 +95,11 @@ def fun_wrapper(*args, **kwargs):
         static = (static_fun,) + static_args_kwargs
         static_leaves, static_treedef = jax.tree_flatten(static)
         static_leaves = tuple(static_leaves)
-        dynamic_out, static_out = _f_wrapped_cache(**jitkwargs)(
+        if isinstance(fun, ft.partial):
+            inner_fun = fun.func
+        else:
+            inner_fun = fun
```

Comment:
So I think this should be `static_fun` (and likewise `static_fun.func` in the partial branch). This is passed into the LRU cache so it needs to be hashable.

By the way, I think this code will fail in the case of passing in `ft.wraps(...)(ft.partial(...))`. This will see the partial and unwrap it, ignoring the fact that it's been `ft.wrap`'d. What do you think the best fix for that is?

Reply:
Why should I use `static_fun` here? Don't we care about the function we're decorating regardless?

Reply:
So the reason is that `inner_fun` is being passed into `_f_wrapped_cache`. This has an LRU cache on it, which demands that all of its arguments be hashable. Mostly that's not an issue -- all functions are hashable -- but we might want to JIT something that isn't just a normal function.

For example a bound method, or a class implementing `__call__`. So we treat `fun` as a PyTree, and split it apart into both dynamic and static parts. Usually both pieces will have the same `__name__` etc. so `ft.wraps` should still just work.

Reply:
In order to handle the `ft.wraps(...)(ft.partial(...))` case I think it'd be best to change the `if` condition to
`if isinstance(fun, ft.partial) and not hasattr(fun, "__name__"):`

Reply:
I had a go at it in 955a7a4, let me know what you think of that solution

Reply:
Ah yeah, that's even better. Nice.

Do you want to bump the version number? We can do a new release for this if this is something that would help you out.

Reply:
Sure, that would be great. Just committed the bump.